src/v/storage/segment.h

```diff
@@ -159,21 +159,33 @@ class segment {
 
     /*
      * Segment Locking
+     * ===============
      *
      * When performing operations on an open segment a lock must be
      * acquired. The appropriate type of lock depends on the operation.
      *
+     * Read locks:
      * Readers should always grab the read lock via `segment::read_lock`.
      * This ensures that no destructive mutations which impact the reader
      * can occur while the lock is held. Multiple callers may hold the read lock
      * at the same time.
      *
+     * Destructive op locks:
      * When performing operations which require total mutual exclusion
      * (e.g. compaction, truncation, etc.) the destructibe lock must be held
      * (acquired via segment::destructive_op_lock). Only one caller may hold
      * this lock at any given time and no read locks may be outstanding.
      *
-     * Lock holders will be granted in the order they were requested.
+     * Read and destructive op lock holders will be granted in the order they
+     * were requested.
+     *
+     * Append locks:
+     * The main purpose of the append lock (`segment::append_lock`) is to
+     * enforce the invariant that only one disk append may be in flight for the
+     * segment at any given time. Internally, `segment` grabs the append lock
+     * every time it performs an append operation. If an external caller holds
+     * the lock that will prevent appends from reaching the disk. Hence, use
+     * with care!
```

Comment:
The append lock is decoupled from the read and destructive operation locks. As a heuristic, callers should minimise the time spent holding the append lock (or even better, not hold the lock at all), since it blocks writes from proceeding.